core/DataAccess/RawLogDao.php

```diff
@@ -328,20 +328,45 @@ protected function insertActionsToKeep($maxIds, $olderThan = true, $insertIntoTe
         $idColumns = $this->getTableIdColumns();
         foreach ($this->dimensionMetadataProvider->getActionReferenceColumnsByTable() as $table => $columns) {
             $idCol = $idColumns[$table];
+            // Create select query for requesting ALL needed fields at once
+            $sql = "SELECT " . implode(',' ,$columns) . " FROM " . Common::prefixTable($table) . " WHERE $idCol >= ? AND $idCol < ?";
 
-            foreach ($columns as $col) {
-                $select = "SELECT $col FROM " . Common::prefixTable($table) . " WHERE $idCol >= ? AND $idCol < ?";
-                $sql = "INSERT IGNORE INTO $tempTableName $select";
-
-                if ($olderThan) {
-                    $start = 0;
-                    $finish = $maxIds[$table];
-                } else {
-                    $start = $maxIds[$table];
-                    $finish = Db::fetchOne("SELECT MAX($idCol) FROM " . Common::prefixTable($table));
+            if ($olderThan) {
+               // Why start on zero? When running for a couple of months, this will generate about 10000+ queries with zero result. Use the lowest value instead.... saves a LOT of waiting time!
+                $start = Db::fetchOne("SELECT MIN($idCol) FROM " . Common::prefixTable($table));;
+                $finish = $maxIds[$table];
+            } else {
+                $start = $maxIds[$table];
+                $finish = Db::fetchOne("SELECT MAX($idCol) FROM " . Common::prefixTable($table));
+            }
+            // Borrowed from Db::segmentedFetchAll
+            // Request records per $insertIntoTempIterationStep amount
+            // Loop over the result set, mapping all numeric fields in a single insert query
+
+            // Insert query would be: INSERT IGNORE INTO [temp_table] VALUES (X),(Y),(Z) depending on the amount of fields requested per row
+            for ($i = $start; $i <= $finish; $i += $insertIntoTempIterationStep) {
+                $currentParams = array($i, $i + $insertIntoTempIterationStep);
+                $result        = Db::fetchAll($sql, $currentParams);
+                // Now we loop over the result set of max $insertIntoTempIterationStep rows and create insert queries
+                $keep_values = [];
+                foreach ($result as $row) {
+                     $keep_values = array_merge($keep_values,array_filter(array_values($row), "is_numeric"));
```

Comment:
is_numeric is important here to avoid SQL injection 👍  maybe we could also add a little comment for this.

Reply:
This was not the first intention, because the data is coming from the database. So there is no sql injection possible. The data that is handled here, should already be checked on SQL injections (by the tracker or log importer)

My idea was that there are null values in the database. So with is_numeric I will filter them out. And that will reduce the amount of inserts that are not needed.